enterprise_access/apps/content_assignments/models.py

```diff
@@ -85,6 +92,137 @@ def policy(self):
         except ObjectDoesNotExist:
             return None
 
+    def _should_acknowledge_expired_assignment(self, assignment):
+        """
+        Returns a tuple of booleans indicating whether the given assignment should be acknowledged and
+        whether the assignment has already been acknowledged.
+        """
+        if assignment.state != LearnerContentAssignmentStateChoices.EXPIRED:
+            return False, False
+
+        last_expiration = assignment.get_last_successful_expiration_action()
+        expiration_last_acknowledged = assignment.get_last_successful_acknowledged_expired_action()
+        if not last_expiration:
+            logger.error(
+                'Assignment %s is in state %s but has no successful expiration action.',
+                assignment.uuid,
+                LearnerContentAssignmentStateChoices.EXPIRED,
+            )
+
+        # Check whether expiration has ever been acknowledged; if not, acknowledge it.
+        if not expiration_last_acknowledged:
+            return True, False
+
+        # If it has been acknowledged before, check whether last expiration action is newer
+        # than the last acknowledged expiration action.
+        if last_expiration and last_expiration.completed_at > expiration_last_acknowledged.completed_at:
+            return True, True
+
+        # Otherwise, the expiration has already been acknowledged and should not be acknowledged again.
+        return False, True
+
+    def _should_acknowledge_cancelled_assignment(self, assignment):
+        """
+        Returns a tuple of booleans indicating whether the given assignment should be acknowledged and
+        whether the assignment has already been acknowledged.
+        """
+        if assignment.state != LearnerContentAssignmentStateChoices.CANCELLED:
+            return False, False
+
+        last_cancellation = assignment.get_last_successful_cancel_action()
+        cancellation_last_acknowledged = assignment.get_last_successful_acknowledged_cancelled_action()
+        if not last_cancellation:
+            logger.error(
+                'Assignment %s is in state %s but has no successful expiration action.',
+                assignment.uuid,
+                LearnerContentAssignmentStateChoices.EXPIRED,
+            )
+
+        # Check whether cancellation has ever been acknowledged; if not, acknowledge it.
+        if not cancellation_last_acknowledged:
+            return True, False
+
+        # If it has been acknowledged before, check whether last cancellation action is newer
+        # than the last acknowledged cancellation action.
+        if last_cancellation and last_cancellation.completed_at > cancellation_last_acknowledged.completed_at:
+            return True, True
+
+        # Otherwise, the cancellation has already been acknowledged and should not be acknowledged again.
+        return False, True
+
+    def acknowledge_assignments(self, assignment_uuids, lms_user_id):
```

Comment:
Thinking about this more, I think it's fairly justifiable for this method to exist in `models.py`.  It's basically saying "hey parent object, go update these specific child objects of yours".  The more complex business logic that crosses domains in `get_subsidy_expiration()` and `get_automatic_expiration_date_and_reason()` feels (to me, at least) as if it belongs in api.py.

Reply:
I agree it would be preferable for `get_automatic_expiration_date_and_reason` and `get_subsidy_expiration` to live within `api.py`. I originally had it there, but then moved it to the models. When trying to replicate the same to see if there was a particular reason, I believe it may have been due to an issue with a circular import:

```shell
  File "/edx/app/enterprise-access/enterprise_access/apps/subsidy_access_policy/models.py", line 18, in <module>
    from enterprise_access.apps.content_assignments import api as assignments_api
  File "/edx/app/enterprise-access/enterprise_access/apps/content_assignments/api.py", line 16, in <module>
    from enterprise_access.apps.content_assignments.tasks import send_reminder_email_for_pending_assignment
  File "/edx/app/enterprise-access/enterprise_access/apps/content_assignments/tasks.py", line 14, in <module>
    from enterprise_access.apps.content_assignments.api import get_automatic_expiration_date_and_reason
ImportError: cannot import name 'get_automatic_expiration_date_and_reason' from partially initialized module 'enterprise_access.apps.content_assignments.api' (most likely due to a circular import) (/edx/app/enterprise-access/enterprise_access/apps/content_assignments/api.py)
```

Maybe we can rubber duck on this for a bit to determine if there's a workaround I'm not seeing to get past this circular import?